File: apps/scanner/src/deepfake/news_search.py
"""News source matching for deepfake investigations.

Uses SerpAPI Google News engine to find news coverage related
to the investigation's subject.
"""

import aiohttp
from sqlalchemy import text

from src.config import settings
from src.db.connection import async_session
from src.deepfake.utils import (
    build_search_query,
    get_investigation_search_context,
    log_activity,
    update_task_progress,
)
from src.utils.logging import get_logger
from src.utils.rate_limiter import get_limiter

log = get_logger("deepfake.news_search")


async def run_news_search(
    task_id: str,
    investigation_id: str,
    parameters: dict,
) -> None:
    """Search Google News for coverage related to the investigation.

    Fetches the investigation's title and geographic context to build
    a search query, then uses SerpAPI's Google News engine.
    """
    if not settings.serpapi_api_key:
        log.warning("serpapi_api_key_not_configured")
        raise ValueError("SerpAPI key not configured — cannot run news search")

    title, geo_context, _ = await get_investigation_search_context(investigation_id)
    query = build_search_query(title, geo_context)

    await update_task_progress(task_id, 20)

    # Call SerpAPI Google News
    limiter = get_limiter("serpapi")
    await limiter.acquire()

    params = {
        "engine": "google",
        "tbm": "nws",
        "q": query,
        "api_key": settings.serpapi_api_key,
        "num": 20,
    }

    async with aiohttp.ClientSession() as http:
        async with http.get("https://serpapi.com/search", params=params) as resp:
            if resp.status != 200:
                body = await resp.text()
                log.warning("serpapi_news_error", status=resp.status, body=body[:500])
                raise RuntimeError(f"SerpAPI news search failed ({resp.status})")

            data = await resp.json()

    await update_task_progress(task_id, 60)

    news_results = data.get("news_results", [])

    # Insert results + log activity
    async with async_session() as session:
        for article in news_results:
            url = article.get("link", "")
            if not url:
                continue
            await session.execute(text(
                "INSERT INTO deepfake_reverse_search_results "
                "(investigation_id, engine, result_url, result_domain, result_title, result_date) "
                "VALUES (:inv_id, 'news_search', :url, :domain, :title, :date)"
            ), {
                "inv_id": investigation_id,
                "url": url,
                "domain": article.get("source", {}).get("name") if isinstance(article.get("source"), dict) else article.get("source"),
                "title": article.get("title"),
                "date": article.get("date"),
            })
        await session.commit()

    await update_task_progress(task_id, 80)

    await log_activity(investigation_id, "news_search_completed", {
        "query": query,
        "results_count": len(news_results),
    })

    log.info(
        "news_search_complete",
        investigation_id=investigation_id,
        query=query,
        results=len(news_results),
    )

File: src/lib/news-search.ts
import { createServiceClient } from "@/lib/supabase/server";
import { logActivity } from "@/lib/investigation-queries";
import { extractDomain } from "@/lib/investigation-utils";

interface SerpApiNewsResult {
  title: string;
  link: string;
  source: string;
  date?: string;
  snippet?: string;
  thumbnail?: string;
}

export async function processNewsSearchTask(investigationId: string): Promise<void> {
  const supabase = await createServiceClient();

  // Claim the pending news_search task
  const { data: tasks, error: taskErr } = await supabase
    .from("deepfake_tasks")
    .select("*")
    .eq("investigation_id", investigationId)
    .eq("task_type", "news_search")
    .in("status", ["pending", "failed"])
    .order("created_at")
    .limit(1);

  if (taskErr) throw taskErr;
  if (!tasks || tasks.length === 0) return;

  const task = tasks[0];

  try {
    await supabase
      .from("deepfake_tasks")
      .update({ status: "running", started_at: new Date().toISOString() })
      .eq("id", task.id);

    const apiKey = process.env.SERPAPI_API_KEY;
    if (!apiKey) {
      await supabase
        .from("deepfake_tasks")
        .update({
          status: "completed",
          completed_at: new Date().toISOString(),
          result: { skipped: true, reason: "SERPAPI_API_KEY not configured" },
        })
        .eq("id", task.id);
      return;
    }

    // Fetch investigation title + region
    const { data: investigation, error: invErr } = await supabase
      .from("deepfake_investigations")
      .select("title, geographic_context")
      .eq("id", investigationId)
      .single();

    if (invErr || !investigation) {
      throw new Error(`Investigation not found: ${investigationId}`);
    }

    // Build search query like the scanner does
    const query = `"${investigation.title}"${investigation.geographic_context ? ` ${investigation.geographic_context}` : ""}`;

    // Call SerpAPI Google News
    const url = new URL("https://serpapi.com/search");
    url.searchParams.set("engine", "google");
    url.searchParams.set("tbm", "nws");
    url.searchParams.set("q", query);
    url.searchParams.set("api_key", apiKey);
    url.searchParams.set("num", "20");

    const res = await fetch(url.toString());
    if (!res.ok) {
      const text = await res.text();
      throw new Error(`SerpAPI error ${res.status}: ${text}`);
    }

    const json = await res.json();
    const newsResults: SerpApiNewsResult[] = json.news_results ?? [];

    // Delete old news_search results before inserting new ones
    await supabase
      .from("deepfake_reverse_search_results")
      .delete()
      .eq("investigation_id", investigationId)
      .eq("engine", "news_search");

    // Insert results, log activity, and mark task complete in parallel
    await Promise.all([
      newsResults.length > 0
        ? supabase
            .from("deepfake_reverse_search_results")
            .insert(
              newsResults.map((r) => ({
                  investigation_id: investigationId,
                  engine: "news_search" as const,
                  result_url: r.link,
                  result_domain: extractDomain(r.link),
                  result_title: r.title,
                  result_date: r.date && !isNaN(Date.parse(r.date)) ? r.date : null,
                  thumbnail_url: r.thumbnail || null,
              }))
            )
            .then(({ error }) => {
              if (error) throw new Error(`Failed to insert news results: ${error.message}`);
            })
        : Promise.resolve(),
      logActivity(investigationId, "news_search_completed", null, {
        query,
        results_count: newsResults.length,
      }),
      supabase
        .from("deepfake_tasks")
        .update({
          status: "completed",
          completed_at: new Date().toISOString(),
          result: { query, results_count: newsResults.length },
        })
        .eq("id", task.id),
    ]);
  } catch (err) {
    const message = err instanceof Error ? err.message : JSON.stringify(err);
    console.error(`[news-search] Task ${task.id} failed:`, message);

    await supabase
      .from("deepfake_tasks")
      .update({
        status: "failed",
        error_message: message,
        retry_count: (task.retry_count || 0) + 1,
        completed_at: new Date().toISOString(),
      })
      .eq("id", task.id);
  }
}
